### ISSUE TITLE
Update SettingRegistry to provide additional information?

### ISSUE THREAD

**Issue by tslaton** [jupyterlab/jupyterlab#6903]
# The problem
I have an extension that writes data to the setting registry programmatically. These programmatic writes trigger the `pluginChanged` signal found [here in the source code](https://github.com/jupyterlab/jupyterlab/blob/master/packages/coreutils/src/settingregistry.ts#L276). (case 1)

If a user directly modifies the extension's settings (eg., as JSON in the settings editor) and saves those modifications, `pluginChanged` is also triggered. (case 2)

As the api is currently written, I don't think there's any way to distinguish whether `pluginChanged` was triggered as a result of a programmatic modification (case 1) or a direct modification (case 2). 

Because I don't want to necessarily wait on a slow network to respond back with a successful save, when modifying the settings programmatically, I would like to reflect the local changes in the extension's settings immediately and quietly save the result to the settings registry in the background. 

However, I'd also like to respond to direct modifications of the settings (case 2) and display the appropriate results. To do that, I need to reload the settings whenever `pluginChanged` is triggered since I can't distinguish the cases. But then I am unnecessarily reloading the settings whenever the reload is the result of a programmatic change I already have accounted for locally - and, what's worse, is that several async changes/reloads could interleave in a way that makes the end result different than expected.

# Proposed solution A
`pluginChanged` could emit not only the name of the plugin changed but also the timestamp of when it was changed. This would be generic and (as far as I know), an easy change. It would solve the problem I'm facing indirectly by allowing me to check the timestamp and only reload if the timestamp of the change is more recent than the working copy of the settings.

I imagine this information would be helpful in handling other scenarios as well.

# Proposed solution B
In addition to `pluginChanged`, there could be an additional signal - maybe called something like `pluginEdited` - that is only emitted when the user directly modifies the settings as JSON in the editor (case 2). Then `pluginChanged` would fire as it currently does in both cases, but the new signal could be used to subscribe/respond to the more specific scenario.

# Code
I can probably implement either and do a PR in the near future. Leaning toward solution A unless someone raises a concern.

**Comment by jasongrout** [comment]
Of course, timestamps can be difficult unless you have a single source of truth clock.

**Comment by telamonian** [comment]
> single source of truth clock

Is that something you can have for a distributed process?

**Comment by tslaton** [comment]
You're right, so you something along the lines of Solution B is going to be better. Solution B also shouldn't break any existing APIs.
